Dockerfile

```diff
@@ -0,0 +1,27 @@
+# Build the manager binary
+FROM golang:1.16 as builder
+
+WORKDIR /workspace
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
+
+# Copy the go source
+COPY main.go main.go
+COPY api/ api/
+COPY controllers/ controllers/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot
+WORKDIR /
+COPY --from=builder /workspace/manager .
+USER 65532:65532
```

Comment:
need to remove user and use the latest image or else we will have GRPC problem see https://github.com/csi-addons/volume-replication-operator/pull/49

Reply:
> need to remove user and use the latest image or else we will have GRPC problem see [csi-addons/volume-replication-operator#49](https://github.com/csi-addons/volume-replication-operator/pull/49)

controller will communicate with sidecar over network. 
For sidecar, we'll have a new build file with `gcr.io/distroless/static:latest` and no user.

Reply:
are we going to have 2 dockerfiles? I thought we are going to have 1 docker image for both controller and sidecar.

Reply:
Have you already verified that with the non-root image you are able to communicate over the wire?

Reply:
> are we going to have 2 dockerfiles? I thought we are going to have 1 docker image for both controller and sidecar.

2 separate images, I'll let @nixpanic confirm. 

> Have you already verified that with the non-root image you are able to communicate over the wire?

root image is required for unix socket communication. 
Controller communicating with k8s api server is over the wire / tcp .
And it being default should work.

Reply:
yes, two container-images. Both contain a different executable, with only the functionality (and permissions) that they need. There is little overlap in functionality between controller and side-car, so it is cleaner to provide images for both functionalities. This also allows security checkers to validate the permissions that an image/executable needs.